File: lib/djv/App/HelpActions.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the DJV project.

#include <djv/App/HelpActions.h>

#include <djv/App/App.h>
#include <djv/App/MainWindow.h>

#include <djv/Models/AppInfoModel.h>

#include <ftk/Core/OS.h>

namespace djv
{
    namespace app
    {
        struct HelpActions::Private
        {
        };

        void HelpActions::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<MainWindow>& mainWindow)
        {
            IActions::_init(context, app, "Help");
            FTK_P();

            std::weak_ptr<MainWindow> mainWindowWeak(mainWindow);

            // Installed beside the application rather than on the web, so
            // that what it describes is the version that is running.
            const std::string docsURL = app->getAppInfoModel()->getDocsURL();

            // Register the commands.
            _addCommand(
                "Documentation",
                "Open the documentation in a web browser.",
                [docsURL](const nlohmann::json&)
                {
                    if (docsURL.empty())
                        return;
                    try
                    {
                        ftk::openURL(docsURL);
                    }
                    catch (const std::exception&)
                    {}
                });

            _addCommand(
                "About",
                "Show the about dialog.",
                [mainWindowWeak](const nlohmann::json&)
                {
                    if (auto mainWindow = mainWindowWeak.lock())
                    {
                        mainWindow->showAboutDialog();
                    }
                });

            _addCommand(
                "SysInfo",
                "Show the system information dialog.",
                [mainWindowWeak](const nlohmann::json&)
                {
                    if (auto mainWindow = mainWindowWeak.lock())
                    {
                        mainWindow->showSysInfoDialog();
                    }
                });

            // Create the actions.
            _actions["Documentation"] = ftk::Action::create(
                "Documentation",
                _command("Documentation"));
            if (docsURL.empty())
            {
                // A build that was not installed has none. Saying so is
                // better than a menu item that does nothing when it is
                // clicked.
                _actions["Documentation"]->setEnabled(false);
                _actions["Documentation"]->setTooltip(
                    "The documentation is installed with the application,\n"
                    "and this build was not installed.");
            }
            _actions["About"] = ftk::Action::create(
                "About",
                _command("About"));
            _actions["SysInfo"] = ftk::Action::create(
                "System Information",
                _command("SysInfo"));
        }

        HelpActions::HelpActions() :
            _p(new Private)
        {}

        HelpActions::~HelpActions()
        {}

        std::shared_ptr<HelpActions> HelpActions::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<MainWindow>& mainWindow)
        {
            auto out = std::shared_ptr<HelpActions>(new HelpActions);
            out->_init(context, app, mainWindow);
            return out;
        }
    }
}

File: examples/python/HelpActions.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the DJV project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl
import djvPy as djv

import IActions
import Util

import weakref

class Actions(IActions.IActions):
    """
    This class provides the help actions.
    """
    def __init__(self, context, app, mainWindow):
        IActions.IActions.__init__(self, context, app, "Help")

        self._app = weakref.ref(app)
        self._mainWindowWeak = weakref.ref(mainWindow)
        self._sysInfoDialog = None

        # Register the commands.
        self._addCommand(
            "Documentation",
            "Open the documentation in a web browser.",
            lambda args, f = Util.weak(self._documentation): f())

        studioURL = app.getAppInfoModel().studioURL
        if studioURL:
            self._addCommand(
                "Studio",
                "Open the DJV Studio web site in a web browser.",
                lambda args: ftk.openURL(studioURL))

        self._addCommand(
            "About",
            "Show the about dialog.",
            lambda args, f = Util.weak(self._about): f())

        self._addCommand(
            "SysInfo",
            "Show the system information dialog.",
            lambda args, f = Util.weak(self._sysInfo): f())

        # Create the actions.
        self.actions["Documentation"] = ftk.Action(
            "Documentation",
            self._command("Documentation"))
        if not app.getAppInfoModel().docsURL:
            # A build that was not installed has none. Saying so is
            # better than a menu item that does nothing when it is
            # clicked.
            self.actions["Documentation"].enabled = False
            self.actions["Documentation"].tooltip = (
                "The documentation is installed with the application,\n"
                "and this build was not installed.")
        if studioURL:
            self.actions["Studio"] = ftk.Action(
                "DJV Studio",
                self._command("Studio"))
            self.actions["Studio"].tooltip = (
                "Open the web site for DJV Studio,\n"
                "the commercial version of DJV.")
        self.actions["About"] = ftk.Action(
            "About",
            self._command("About"))
        self.actions["SysInfo"] = ftk.Action(
            "System Information",
            self._command("SysInfo"))

        # Register the shortcuts.
        self._addShortcut("Documentation")
        self._addShortcut("About")
        self._addShortcut("SysInfo", "System Information")

        self._shortcutsUpdate(self._settingsModel.shortcuts)

    def _documentation(self):
        info = self._app().getAppInfoModel()
        ftk.openURL(info.docsURL)

    def _about(self):
        info = self._app().getAppInfoModel()
        text = "{} version: {}\nCommit date: {}\nCommit: {}".format(
            info.fullName,
            info.version,
            info.commitDate,
            info.gitCommit)
        dialogSystem = self._app().context.getSystemByName("ftk::DialogSystem")
        dialogSystem.message("About", text, self._mainWindowWeak())

    def _sysInfo(self):
        app = self._app()
        mainWindow = self._mainWindowWeak()
        text = djv.ui.getSysInfo(
            app.context,
            app.getAppInfoModel(),
            app.getSettingsModel(),
            mainWindow.getWindowInfo())
        selfWeak = weakref.ref(self)
        self._sysInfoDialog = djv.ui.SysInfoDialog(app.context, text)
        self._sysInfoDialog.open(mainWindow)
        self._sysInfoDialog.setCloseCallback(
            lambda: selfWeak() and setattr(
                selfWeak(), "_sysInfoDialog", None))
